src/imitation/algorithms/bc.py

```diff
@@ -357,48 +397,57 @@ def train(
                 even if `.train()` logged to Tensorboard previously. Has no practical
                 effect if `.train()` is being called for the first time.
         """
-        it = EpochOrBatchIteratorWithProgress(
-            self._demo_data_loader,
-            n_epochs=n_epochs,
-            n_batches=n_batches,
-            on_epoch_end=on_epoch_end,
-            on_batch_end=on_batch_end,
-            progress_bar_visible=progress_bar,
+        if reset_tensorboard:
+            self._bc_logger.reset_tensorboard_steps()
+
+        compute_rollout_stats = RolloutStatsComputer(
+            log_rollouts_venv,
+            log_rollouts_n_episodes,
         )
 
-        if reset_tensorboard:
-            self.tensorboard_step = 0
+        def _on_epoch_end(epoch_number: int):
+            if isinstance(batches_with_stats, tqdm.tqdm):
```

Comment:
The need for an `isinstance` is a little sad but probably harmless, does mean we can't wrap `batches_with_stats` any more without it causing this check to fail.

One alternative would be to have `batches_with_progress_bar` initialized to `None` at same time batches_with_stats` is set, and then `if progress_bar:` update both `batches_with_stats` and `batches_with_progress_bar` to tqdm wrapped version. Then we can drop line 411 that does a type cast, and replace the `isinstance` with just `if batches_with_progress_bar is not None`. WDYT?

Reply:
Its more or less the same LOC and equally confusing but I like that it is robust to further wrapping of the iterators. I implemented it!